tests/test_analytical_simulator.py

```diff
@@ -0,0 +1,186 @@
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+
+import unittest
+import numpy as np
+
+from openqaoa.backends.simulators.qaoa_analytical_sim import QAOABackendAnalyticalSimulator 
+#from openqaoa.utilities import random_k_regular_graph
+from openqaoa.workflows.optimizer import QAOA, RQAOA
+from openqaoa.problems.problem import QUBO, MaximumCut
+from openqaoa.devices import create_device
+from openqaoa.utilities import X_mixer_hamiltonian, ring_of_disagrees, random_k_regular_graph
+from openqaoa.qaoa_parameters.baseparams import QAOACircuitParams
+from openqaoa.qaoa_parameters import QAOAVariationalStandardParams
+
+"""
+A set of tests for the analytical simulator backend which computes the energy of a given quantum circuit as a function of beta and gamma.
+"""
+def Disagrees_SetUp(n_qubits, p, mixer_hamil):
+    """
+    Helper function for the tests below
+    """
+
+    register = range(n_qubits)
+    cost_hamil = ring_of_disagrees(register)
+    
+    betas = [np.pi/8]
+    gammas = [np.pi/4]
+
+    qaoa_circuit_params = QAOACircuitParams(cost_hamil, mixer_hamil, p)
+    variational_params_std = QAOAVariationalStandardParams(qaoa_circuit_params, betas, gammas)
+    # Get the part of the Hamiltonian proportional to the identity
+
+    return register, cost_hamil, qaoa_circuit_params, variational_params_std
+
+
+class TestingQAOABackendAnalyticalSimulator(unittest.TestCase):
+    """
+    Unittest based testing of QAOABackendAnalyticalSimulator
+    """
+
+    def test_expectation(self):
+        """
+        Testing if the analytical formula returns the energy expectation value for a given beta and gamma on an easy problem (Ring of disagree). 
+        """
+        n_qubits = 8
+        p = 1
+        mixer_hamil = X_mixer_hamiltonian(n_qubits)
+        register, cost_hamil, qaoa_circuit_params, variate_params = Disagrees_SetUp(n_qubits, p, mixer_hamil)
+
+        backend_analytical = QAOABackendAnalyticalSimulator(qaoa_circuit_params)
+        exp_val = backend_analytical.expectation(variate_params)
+
+        # Check correct expecation value
+        assert np.isclose(exp_val, -6)
+        
+    def test_p_not_1_fails(self):
+        """
+        Testing if the analytical backend fails if the number of layers, p, is different than 1. 
+        """
+        exception = False
+        try:
+            n_qubits = 8
+            p = 2
+            mixer_hamil = X_mixer_hamiltonian(n_qubits)
+            register, cost_hamil, qaoa_circuit_params, variate_params = Disagrees_SetUp(n_qubits, p, mixer_hamil)
+            backend_analytical = QAOABackendAnalyticalSimulator(qaoa_circuit_params)
```

Comment:
Codecov says that this line has never been tested: this seems to imply that the test failed BEFORE reaching line 79. Is this the desired behaviour?

Reply:
Ah, good catch! It was failing before that indeed. should be fixed now!